Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewCell.cs

```diff
@@ -0,0 +1,1265 @@
+        // protected virtual object GetFormattedValue(object value, int rowIndex, ref DataGridViewCellStyle cellStyle, TypeConverter valueTypeConverter, TypeConverter formattedValueTypeConverter, DataGridViewDataErrorContexts context)
+
+        /// <summary>
+        /// Gets the inherited shortcut menu for the current cell.
+        /// </summary>
+        /// 
+        /// <returns>
+        /// A <see cref="T:System.Windows.Forms.ContextMenuStrip"/> if the parent <see cref="T:System.Windows.Forms.DataGridView"/>, <see cref="T:System.Windows.Forms.DataGridViewRow"/>, or <see cref="T:System.Windows.Forms.DataGridViewColumn"/> has a <see cref="T:System.Windows.Forms.ContextMenuStrip"/> assigned; otherwise, null.
+        /// </returns>
+        /// <param name="rowIndex">The row index of the current cell.</param><exception cref="T:System.ArgumentOutOfRangeException">The value of the <see cref="P:System.Windows.Forms.DataGridViewElement.DataGridView"/> property of the cell is not null and the specified <paramref name="rowIndex"/> is less than 0 or greater than the number of rows in the control minus 1. </exception><exception cref="T:System.InvalidOperationException"><see cref="P:System.Windows.Forms.DataGridViewCell.ColumnIndex"/> is less than 0, indicating that the cell is a row header cell.</exception>
+        public virtual ContextMenuStrip GetInheritedContextMenuStrip(int rowIndex)
+        {
+            Contract.Requires(this.DataGridView != null);
+            Contract.Requires(rowIndex >= 0 || rowIndex < this.DataGridView.Rows.Count);
+            Contract.Requires(this.ColumnIndex < 0);
+            return default(ContextMenuStrip);
+        }
+
+        /// <summary>
+        /// Returns a value indicating the current state of the cell as inherited from the state of its row and column.
+        /// </summary>
+        /// 
+        /// <returns>
+        /// A bitwise combination of <see cref="T:System.Windows.Forms.DataGridViewElementStates"/> values representing the current state of the cell.
+        /// </returns>
+        /// <param name="rowIndex">The index of the row containing the cell.</param><exception cref="T:System.ArgumentException">The cell is not contained within a <see cref="T:System.Windows.Forms.DataGridView"/> control and <paramref name="rowIndex"/> is not -1.-or-<paramref name="rowIndex"/> is not the index of the row containing this cell.</exception><exception cref="T:System.ArgumentOutOfRangeException">The cell is contained within a <see cref="T:System.Windows.Forms.DataGridView"/> control and <paramref name="rowIndex"/> is outside the valid range of 0 to the number of rows in the control minus 1.</exception>
+        public virtual DataGridViewElementStates GetInheritedState(int rowIndex)
+        {
+            Contract.Requires(this.DataGridView != null);
```

Comment:
This one is not correct as well (see http://referencesource.microsoft.com/#System.Windows.Forms/winforms/Managed/System/WinForms/DataGridViewCell.cs,8bf37b550e753667).

Should be something like this:

`Contract.Requires(DataGridView == null || rowIndex == -1, "When Grid is null, index should be -1")`
`Contract.Requires(DataGridView != null || (rowIndex >= 0 && rowIndex < DataGridView.Rows.Count);`